espnet3/collect_stats.py

```diff
@@ -1,16 +1,435 @@
+        )
+
+
+def _collect_stats_common(
+    *,
+    model_config,
+    dataset_config,
+    dataloader_config,
+    mode: str,
+    output_dir: Path,
+    task: Optional[str],
+    write_collected_feats: bool,
+    batch_size: int,
+    shard_idx: Optional[int] = None,
+    shape_key_suffix: str = "",
+    sum_dict: Optional[Dict] = None,
+    sq_dict: Optional[Dict] = None,
+    count_dict: Optional[Dict] = None,
+    writers: Optional[Dict] = None,
+):
+    num_items = _dataset_length(dataset_config, mode, shard_idx)
+    index_batches = _chunk_indices(num_items, batch_size) if num_items else []
+
+    provider = CollectStatsInferenceProvider(
+        model_config=model_config,
+        dataset_config=dataset_config,
+        dataloader_config=dataloader_config,
+        mode=mode,
+        task=task,
+        shard_idx=shard_idx,
+        params={"write_collected_feats": write_collected_feats},
+    )
+    runner = CollectStatsRunner(provider)
+
+    sum_dict = sum_dict or defaultdict(lambda: 0)
+    sq_dict = sq_dict or defaultdict(lambda: 0)
+    count_dict = count_dict or defaultdict(lambda: 0)
+    writers = writers or {}
+
+    results = runner(index_batches)
+
+    with DatadirWriter(output_dir / mode) as datadir_writer:
+        for result in results:
+            if write_collected_feats:
+                stats, shape_info, feats = result
+            else:
+                stats, shape_info = result
+                feats = None
+
+            _accumulate_and_persist_batch(
+                stats=stats,
+                shape_info=shape_info,
+                feats=feats,
+                sum_dict=sum_dict,
+                sq_dict=sq_dict,
+                count_dict=count_dict,
+                datadir_writer=datadir_writer,
+                writers=writers,
+                mode=mode,
+                output_dir=output_dir,
+                write_collected_feats=write_collected_feats,
+                shape_key_suffix=shape_key_suffix,
+            )
+
+    return sum_dict, sq_dict, count_dict
+
+
+def collect_stats_multiple_iterator(
+    model_config,
+    dataset_config,
+    dataloader_config,
+    mode: str,
+    output_dir: Path,
+    task: Optional[str],
+    batch_size: int,
+    parallel_config,
+):
+    """Collect stats on sharded datasets using Dask + setup_fn."""
+    set_parallel(parallel_config)
+
+    sum_dict, sq_dict, count_dict = (
+        defaultdict(lambda: 0),
+        defaultdict(lambda: 0),
+        defaultdict(lambda: 0),
+    )
+
+    mode_cfg = getattr(dataloader_config, mode)
+    num_shards = mode_cfg.num_shards
+
+    for shard_idx in range(num_shards):
+        sum_dict, sq_dict, count_dict = _collect_stats_common(
+            model_config=model_config,
+            dataset_config=dataset_config,
+            dataloader_config=dataloader_config,
+            mode=mode,
+            output_dir=output_dir,
+            task=task,
+            write_collected_feats=False,
```

Comment:
Why is write_collected_feats set to `False` here?

Reply:
Oh, thank you maybe this is from debugging  stage...?
I also noticed that I didn't allowed this flag with multiple iterator (which should be supported). So I fixed this part